Forbidden recursive destruct for beds' parts

LeMagnesium opened the pull request:
- Whenever a part of a bed is destroyed, it uses a metadata flag to indicate
   to the second part not to destroy it again

EDIT: Solves #513 for beds. See my comment there about doors.

HybridDog commented on mods/beds/api.lua:
```diff
@@ -97,6 +99,20 @@ function beds.register_bed(name, def)
 			type = "fixed",
 			fixed = {0, 0, 0, 0, 0, 0},
 		},
+		on_destruct = function(pos)
```
Are there two "on_destruct"s?

LeMagnesium replied:
Yes. Both nodes have a destruct function. If the top is destroyed, then it will trigger the bottom's `on_destruct` and will 'tell' it not to care about destroying the first node. Following the same system, if the bottom node is destroyed, it will trigger the top's `on_destruct` callback, which won't destroy recursively the first node.

HybridDog replied:
l added the on_destruct to only one part of the bed.
https://github.com/minetest/minetest_game/pull/488/files#diff-b2333dc7fcdfecb27e6bed8daa7d3af8L40
Maybe it's faster if you use a local variable instead of the meta.

LeMagnesium replied:
Adding it to only one part of the bad makes it very complicated (I will not say impossible, but not worth the complexity of the needed process) to make the second part break with the first one, we need a callback which can apply from both of the nodes, and here, there are two `on_destruct`, one per part.
Then, using a local variable might be faster, but more complicated. The only way I see is to create a local table, whose indexes are positions of dug part of beds, and values are always `1`, or something like `top` or `bottom`, depending on the destroyed part. In either way, it's more complicated than `minetest.get_meta(pos):[get/set]_int` which is an already implemented system to store datas related to a node, per node.

HybridDog commented:
>Then, using a local variable might be faster, but more complicated.

ln my opinion it's very easy. l think you could use a boolean. If it's false, set it to true and remove the other bed, and if it's true, set it to false and abort.

LeMagnesium commented:
> ln my opinion it's very easy. l think you could use a boolean. If it's false, set it to true and remove the other bed, and if it's true, set it to false and abort.

Wait, a boolean? For every bed on a map? What if two people break their beds at the same time then?

HybridDog commented:
> What if two people break their beds at the same time then?

that's impossible

LeMagnesium commented:
Well, if it can't happen, then I will rewrite it.

LeMagnesium commented:
Rewritten, using a local variable in `mods/beds/api.lua` .

HybridDog commented:
l don't think it works this way, if the top part is removed once, do_not_destruct is always true.
l think you would need to do the same changes on both parts:
At first, before defining n, l would write
> if do_not_destruct then
do_not_destruct = false
return
end

Before calling minetest.remove_node l would write
>do_not_destruct = true

HybridDog commented on mods/beds/api.lua:
```diff
@@ -97,6 +100,19 @@ function beds.register_bed(name, def)
+			local dir = minetest.facedir_to_dir(n.param2)
+			-- Reverse direction
+			dir = {x = -dir.x, y = -dir.y, z = -dir.z}
+			local p = vector.add(pos,dir)
```
for the reverse direction you could also use vector.subtract(pos,dir)

LeMagnesium replied:
Thanks for this method, I'm not very used to vector functions.
`do_no_destruct` and `vector.substract` will be fixed and pushed in about minutes.

LeMagnesium commented:
I fixed the mistake about `do_not_destruct`, and added `vector.subtract` . Now I will follow your advise and create a local function. You can comment on the content of the functions, I will keep it as it is now, and just add an `if` statement.

asl97 commented:
was going to say how it was doing:

1. player dig bottom
2. bottom remove top
3. top remove bottom (again) because it doesn't check the variable

____
using a public function not only make things simpler and much more clear,
~~there could be some micro optimization but it's negligible and is just the perfectionist side of me talking~~

LeMagnesium commented:
Maybe I introduced a bug with the two `on_destruct`, althrought I have tested both top and bottom dig and both worked without recursion. Then, the public function must have solved this, because it checks the variable in the if statement checking for the detected node to be a bed of the good group and correctly orientated.

paramat commented:
Bump, needs review.

sofar commented:
A much easier solution would be to make the top (unreachable) part of the bed non-flammable, non-diggable, non-explodable, and not have an on_destruct() method. This is how I solved this in doors. 

Chances that the top part of a bed sits in a fire but the bottom doesn't are very, very slim.

LeMagnesium commented:
> A much easier solution would be to make the top (unreachable) part of the bed non-flammable, non-diggable, non-explodable, and not have an on_destruct() method

I'm going to have to rebase anyway.. So I'll do that and try that method.

Ekdohibs commented:
My solution would be to make some code to handle such multi-nodes situations, and use it for beds, doors, etc.

LeMagnesium commented:
There, it's rebased. Now I'll change some code, make tests, and let you know when it's okay if it's okay.

> My solution would be to make some code to handle such multi-nodes situations, and use it for beds, doors, etc.

That would be great, but good luck to get that merged with everyone's approval.

LeMagnesium commented:
Ok, after many tests, I've concluded that @sofar 's solution is easier and works.
There's one case though that doesn't work : when using @VanessaE 's pipework nodebreakers

![](https://lut.im/SomMfMQk0X/26r3vmIeLVTxDmnK)

I'm going to keep the destroy functions as I previously coded them, but bed tops are now non-flammable and non-diggable. I think pipeworks should be somehow changed so that nodebreakers can determine that even though they're pointing at the top node it should destroy to bottom part (which sounds like a pain to implement).

sofar commented:
The nodebreaker node should not be able to break nodes that are not diggable. That would be a better solution.

sofar commented:
I really don't like the use of the local ```do_not_destroy``` variable. The ```on_destruct()``` for the top half should just not even have an ```on_destruct```. The ```on_destruct()``` for the bottom bed part should remove the top part, always.

https://github.com/minetest/minetest_game/blob/master/mods/doors/init.lua#L75

just do it like doors does.

paramat commented:
^ What sofar suggests.

LeMagnesium commented:
> I really don't like the use of the local do_not_destroy variable. The on_destruct() for the top half should just not even have an on_destruct. The on_destruct() for the bottom bed part should remove the top part, always.

It's not just as though I had explained above... well maybe it wasn't clear. I would gladly remove that ugly variable and the destroy function, but some mods do not care about poiting at nodes and remove the top part **see my comment above with the screenshot** and you cannot ommit the case where something removes the top part, even when it's not flammable nor pointable, it can still remove it and the other part of the bed will stay.

> just do it like doors does.

Fine! Let's use nodemeshes! We won't have to deal with double nodes managing cr*p.

I'm going to do exactly what you tell me and purposely let the case of something digging the top part unhandled. When someone comes to you and shows you a screenshot of a bed half destroyed by a node breaker, just give them the link to this issue.

paramat commented:
Ok i should stay neutral as i don't yet fully understand.

sofar commented:
The problem is that mods shouldn't destroy certain nodes, especially nodes that have these properties:

```
	pointable = false,
	diggable = false,
	floodable = false,
```

The nodebreaker node especially should test for nodes being ```diggable``` and not touch nodes that are not ```diggable```.

As a matter of fact, the nodebreaker should always just use ```minetest.node_dig(pos, ..)``` to do it's work.

The bug is entirely in that mod. And it's easily fixable.

```minetest_game``` by itself should be coherent. So if flames or TNT only half-destroy a bed, then there is a bug.

In doors, this is solved by

1) not making door tops ```flammable```
2) door tops are not ```diggable```
3) ```on_blast = function() end``` so TNT can't break them.

this removes the need for door tops to even have an ```on_destruct()```.

> Fine! Let's use nodemeshes! We won't have to deal with double nodes managing cr*p.

haha, yes, I totally get your frustration. I would love to get rid of the simple bed myself. But even with a fancy bed, you need to prevent block placement on the node half.

sofar commented:
And looking at this code (assuming this is the actual nodebreaker we're talking about here):

https://github.com/VanessaE/pipeworks/blob/af2c3778feb73ced41c63e3fb8e6bfafdde03bf9/wielder.lua#L356

it's actually using minetest.node_dig.... :)

So, make the bed top ```diggable = false``` and ```flammable = false ``` and ```floodable = false``` and you can remove the entire ```on_destruct``` from the top, and let the bed-bottom on_destruct() do all the work.

paramat commented:
Problem is there are ways that half a bed can be removed by a voxelmanip which does not call any on_destruct callbacks, so both parts of the bed should always be diggable otherwise there is a chance of an un-removable node in the world.
So i would say make both parts of the bed diggable and flammable and then this PR seems okay as it is.

HybridDog commented:
https://github.com/minetest/minetest_game/pull/488/files

LeMagnesium commented:
> So, make the bed top diggable = false and flammable = false and floodable = false and you can remove the entire on_destruct from the top, and let the bed-bottom on_destruct() do all the work.

Yet :

> So i would say make both parts of the bed diggable and flammable and then this PR seems okay as it is.

Please agree together on one and I'll implement it. I can't have contradicting advices and pick who I'm going to stand against.

sofar commented:
Right, I had given no thought to VoxelManip.

I have my doubt that VoxelManips are a good argument. Even VoxelManip code in mods should honor node properties and prevent digging nodes that are not diggable. If there are mods that use VoxelManips that erase nodes, does it even call on_destruct()? There are so many uncertainties that I have my doubts that this is a valid argument. Or, maybe there is some valid VoxelManip code in core that does this? Any examples?

For TNT vmanips we honor all the nodes' properties. I don't see that it be otherwise.

If VoxelManips are indeed a good argument, then I'd avoid using a local variable by coding it up like this:

top bed part:
```
    on_destruct = function(pos)
        beds.destroy(pos, 2)
    end,
```

bottom bed part:

```
    on_destruct = function(pos)
        beds.destroy(pos, 1)
    end,
```

```
function beds.destroy(pos, n)
    if n == 2 then
        other = -- calc bottom
    elseif n == 1 then
        other = -- calc top
    end

    if n < 2 then
        minetest.remove_node(pos)
    end
    if n > 0    
        beds.destroy(other, n - 1)
    end
end
````

while lua in minetest is single-threaded, using a larger-scope variable to keep state is a sure way to get all messed up, and an independent implementation like above avoids that problem is is probably easier to debug.

sofar commented:
@paramat does sapling growth use vmanips that grow trunks through beds?

paramat commented:
Voxelmanips do not trigger callbacks, that's what makes them fast. Most mod uses of voxelmanip don't add code to call callbacks, so yes i'm concerned about mod voxelmanips.
Growing saplings don't force-place any nodes, that's why we have different tree schematics for growing saplings.
So it quite likely a mod voxelmanip will remove half a bed.
LeMagnesium i'm a dev so you can listen to me hehe, leave the PR as it is for now it's probably okay.

sofar commented:
Both parts of the bed are already diggable.

I've been banging my head against this now for an hour, and the last patch doesn't seem right either (```reverse``` isn't ever read).

I think the problem is that calling ```remove_node()``` is just calling the next ```on_destruct``` and queuing the map updates but not actually doing this. So if you call ```get_node()``` inside an ```on_destruct()``` it's unreliable.

*ponders*

sofar commented:
I'm totally sorry I didn't actually look at the underlying cause and effect before reviewing this patch. @LeMagnesium actually had working code in several of the revisions, and I made him break it :(

I've pushed a fixed up version that addresses the issue and technically describes the disaster that is happening in this case, and why it's the best fix I can come up with as well. Please review. Thanks.

paramat commented:
#911 is approved.

PilzAdam commented:
> #911 is approved.

@paramat confirmed terrorist.

paramat commented:
lol
